src/FilePathsBase.jl

```diff
@@ -69,33 +69,40 @@ function __init__()
     register(WindowsPath)
 end
 
-abstract type AbstractPath <: AbstractString end
+"""
+    AbstractPath
+
+Defines an abstract filesystem path. Subtypes of `AbstractPath` should implement the
+following methods:
+
+- `Base.String(p)`
+- `FilePathsBase.parts(p)`
+- `FilePathsBase.root(p)`
+- `FilePathsBase.drive(p)`
+- `FilePathsBase.ispathtype(::Type{MyPath}, x::AbstractString) = true`
+"""
+abstract type AbstractPath end
 
 function register(T::Type{<:AbstractPath})
-    # We add the type to the beginning of our PATH_TYPES, 
-    # so that they can take precedence over the Posix and 
+    # We add the type to the beginning of our PATH_TYPES,
+    # so that they can take precedence over the Posix and
     # Windows paths.
     Compat.pushfirst!(PATH_TYPES, T)
 end
 
-# Required methods for subtype of AbstractString
-Compat.lastindex(p::AbstractPath) = lastindex(String(p))
-Compat.ncodeunits(p::AbstractPath) = ncodeunits(String(p))
-if VERSION >= v"0.7-"
-    Base.iterate(p::AbstractPath) = iterate(String(p))
-    Base.iterate(p::AbstractPath, state::Int) = iterate(String(p), state)
-else
-    Base.next(p::AbstractPath, i::Int) = next(String(p), i)
+#=
+We only want to print the macro string syntax when compact is true and
+we want print to just return the string (this allows `string` to work normally)
+=#
+Base.print(io::IO, path::AbstractPath) = print(io, String(path))
+function Base.show(io::IO, path::AbstractPath)
+    get(io, :compact, false) ? print(io, path) : print(io, "p\"$(String(path))\"")
 end
 
-# The following should be implemented in the concrete types
-Base.String(path::AbstractPath) = error("`String not implemented")
-parts(path::AbstractPath) = error("`parts` not implemented.")
-root(path::AbstractPath) = error("`root` not implemented.")
-drive(path::AbstractPath) = error("`drive` not implemented.")
-
 Base.convert(::Type{AbstractPath}, x::AbstractString) = Path(x)
-ispathtype(::Type{T}, x::AbstractString) where {T<:AbstractPath} = false
+Base.convert(::Type{String}, x::AbstractPath) = string(x)
```

Comment:
This should be `String(x)` not `string(x)`.

Both semantically, and because `string(x)` gives the wrong result.

```
julia> pp=Path(pwd())
p"//Users/oxinabox/JuliaEnvs/MagneticReadHead"

julia> string(pp)
p"//Users/oxinabox/JuliaEnvs/MagneticReadHead"

julia> String(pp)
"/Users/oxinabox/JuliaEnvs/MagneticReadHead"
```

Reply:
That's changing with the `show` and `print` changes.

```
julia> using FilePathsBase
[ Info: Recompiling stale cache file /Users/rory/.playground/share/filepaths/depot/compiled/v1.0/FilePathsBase/Efv1v.ji for FilePathsBase [48062228-2e41-5def-b9a4-89aafe57970f]

julia> p = cwd()
p"//Users/rory/repos/FilePathsBase.jl"

julia> string(p)
"/Users/rory/repos/FilePathsBase.jl"

julia> String(p)
"/Users/rory/repos/FilePathsBase.jl"
```

While this achieves what I'd like, I'm not sure it's the correct way to handle that.

Reply:
`String(p)` feels more semantic to me.
It feels like a shorthand for `convert(String, p)`

Where as `string(p)` to me reads like "Give me some thing I could show a user".
Since it is tied to just being `print`.

Further, I think some `AbstractPaths` may want to overload their `print` in other ways.
Since `print` is pretty freefrom in what you do. 
Like an `S3Path` might want to actually display the URL rather than (just) the path.